### ISSUE TITLE
Use of #line directives in the output file

### ISSUE THREAD

**Issue by simark** [swig/swig#1591]
The C code we write between `%{ .. %}`, ends up in the big C file generated by SWIG.  The errors reported by gcc are then in term of this file.  We have to manually map back to the original line in the SWIG file.

I am sure I am not the first one to think about that, so maybe I am missing something obvious, but is there a reason SWIG does not put `#line` directives in its output?

Ref: https://gcc.gnu.org/onlinedocs/cpp/Line-Control.html#Line-Control

**Comment by vadz** [comment]
FWIW this looks like a good idea to me, I think the reason it hasn't been done yet is just that nobody did it.

**Comment by simark** [comment]
Ok, thanks for the answer.  I took a look at the swig code, see if I could manage to add it, but I quickly realized it would be a bit too complex for somebody without knowledge of the code base.

**Comment by simark** [comment]
For anybody stumbling on this, here's an easy alternative top get actual line numbers in compiler error messages.  Instead of writing code directly in `%{...%}`, place it in another file that you include:

```
%{
#include "my-code.h"
%}
```

The error messages will go from something like:

```
bt2/native_bt.c:3362:8: error: expected declaration specifiers or ‘...’ before numeric constant
 3362 | printf(2);
      |        ^
```

to

```
/home/simark/src/babeltrace/src/bindings/python/bt2/bt2/native_bt_component_class.i.h:26:8: error: expected declaration specifiers or ‘...’ before numeric constant
   26 | printf(2);
      |        ^
```

It will also likely be more comfortable, as your editor will know it's C code.

**Comment by ojwb** [comment]
This was originally suggested over 20 years ago and Dave rejected it having tried it previously: https://sourceforge.net/p/swig/bugs/95/

There's was a newer (but still ancient) ticket I recent closed as well which also has background: https://sourceforge.net/p/swig/feature-requests/40/

(I did try to check for relevant tickets here when I closed the old one but github's ticket search unhelpfully ignores the `#` and a search for `#line` returns over 100 matches, most irrelevant.)

I'll repeat my closing comment from that less-old ticket here:

> That other ticket says:
>
>>    I tried this once before in a very early SWIG release. It
    had exactly the opposite effect---instead of making the
    wrapper files easier to debug, it made them nearly
    impossible to debug. This was mainly due to the fact that
    there is really not a 1-1 correspondence between the
    structure of a .i file and the resulting _wrap.c file (i.e.,
    each declaration in the wrap file may expand to dozens of
    lines of wrapper code). Because of this, it's very
    difficult to get line numbers to track correctly (unless you
    put a #line directive on every other line of the wrapper
    file). More often than not, I found that error messages
    were reported for random line numbers that made no sense
    whatsoever.
>
>    If we're going to do anything here, I am much more inclined to
    stick descriptive comments in the wrapper file that point
    back to a specific .i file location. Past experience
    with #line was such a disaster I don't think I want to
    try that again.
>
>I can see Dave's points and they generally seem valid, so on the basis of that and since we're now 11 years on and nobody's come up with a patch of any sort I'm going to close this as "wont-fix" too. I think a key point is that if the #line generation isn't perfect then it will often be more harmful than helpful.
>
>It can certainly be annoying though when there's a big chunk of verbatim text from another file and the error just points to the generated file. Perhaps #line for %inline and similar but not for expanded typemaps would be more help than hindrance?
>
>If someone thinks they can improve the situation and actually wants to work on a patch, please open a PR at https://github.com/swig/swig

**Comment by simark** [comment]
Yeah, my original request was just the the user code between `%{ ... %}`, which appears verbatim in the generated file.

I don't use touch the SWIG files that much nowadays, that part of our project is "done" and works well.  As mentioned above, we used the "include" alternative to work around this:

https://github.com/efficios/babeltrace/blob/2bf8d3246aba9009042c293bb0154c7738e47ae3/src/bindings/python/bt2/bt2/native_bt_graph.i#L91

I don't expect `#line` to be used for all the rest, the code that is generated from the function declarations.

**Comment by ojwb** [comment]
Probably `%inline` and `%insert` as well as `%{ ... %}`.

It should probably be possible to disable too.

One awkwardness with implementing is there isn't a way to revert to the natural line numbering of the file, so you need to keep count of what that would be to use in `#line 1234 foo_wrap.cxx`. That means any inserted text will need scanning to count the number of newlines.  Not too hard as such, but retrofitting may be painful as it needs to reliably happen in every place that inserts text or else line numbers can get out of whack.

**Comment by simark** [comment]
> One awkwardness with implementing is there isn't a way to revert to the natural line numbering of the file, so you need to keep count of what that would be to use in `#line 1234 foo_wrap.cxx`. That means any inserted text will need scanning to count the number of newlines. Not too hard as such, but retrofitting may be painful as it needs to reliably happen in every place that inserts text or else line numbers can get out of whack.

Good observation.  I thought there would be some kind of `#line pop` directive, to come back to the "natural" numbering, but no.
